File: scouter.webapp/src/main/java/scouterx/webapp/framework/cache/XLogLoopCache.java
/*
 *  Copyright 2015 the original author or authors.
 *  @https://github.com/scouter-project/scouter
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

package scouterx.webapp.framework.cache;

import lombok.extern.slf4j.Slf4j;
import scouter.lang.pack.XLogPack;
import scouter.util.IntSet;
import scouterx.webapp.framework.client.server.Server;
import scouterx.webapp.model.XLogPackWrapper;
import scouterx.webapp.framework.configure.ConfigureAdaptor;
import scouterx.webapp.framework.configure.ConfigureManager;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;

/**
 * Cache of latest XLog data
 *
 * @author Gun Lee (gunlee01@gmail.com) on 2017. 9. 9.
 */
@Slf4j
public class XLogLoopCache {
    private static final Map<Integer, XLogLoopCache> loopCacheMap = new HashMap<>();

    private static ConfigureAdaptor conf = ConfigureManager.getConfigure();
    private final int MAX_RETRIEVE_COUNT;
    private final Server server;
    private long loop = 0;
    private int index = 0;
    private final XLogPackWrapper[] queue;

    public XLogLoopCache(Server server, int capacity) {
        this.MAX_RETRIEVE_COUNT = capacity / 2;
        this.server = server;
        if (server == null) {
            throw new RuntimeException("Not yet initialized!");
        }
        queue = new XLogPackWrapper[capacity];
        loopCacheMap.put(server.getId(), this);
    }

    public static XLogLoopCache getOf(int serverId) {
        return loopCacheMap.get(serverId);
    }

    public void add(XLogPack pack) {
        if (conf.isTrace() && index % 10000 == 0) {
            log.info("XLog added - {} {} {}", loop, index, pack);
        }

        synchronized (queue) {
            queue[index] = new XLogPackWrapper(pack, this.loop, this.index);
            index += 1;
            if (index >= queue.length) {
                loop += 1;
                index = 0;
            }
        }
    }

    public void getAndHandleRealTimeXLog(long aLoop, int aIndex, int aMaxCount, long bufferTime, Consumer<XLogPackWrapper> handlerConsumer) {
        getAndHandleRealTimeXLog(null, aLoop, aIndex, aMaxCount, bufferTime, handlerConsumer);
    }

    /**
     * get new xlog data revealed after the previous check point, and invoke the handler consumer
     *
     * @param aLoop - last loop retrieved
     * @param aIndex - last index retrieved
     * @param bufferTime - buffer time(milliseconds) for to prevent fetching data before dictionary data generation.
     */
    public void getAndHandleRealTimeXLog(IntSet objHashIntSet, long aLoop, int aIndex, int aMaxCount, long bufferTime,
                                         Consumer<XLogPackWrapper> handlerConsumer) {
        int maxCount = aMaxCount;

        //Initial call
        if(aLoop == 0 && aIndex == 0) {
        } else if(++aIndex >= queue.length) {
            aIndex = 0;
            aLoop++;
        }

        long currentLoop;
        int currentIndex;
        synchronized (queue) {
            currentLoop = this.loop;
            currentIndex = this.index;
        }

        int loopStatus = (int) (currentLoop - aLoop);

        switch (loopStatus) {
            case 0:
                int countToGet = currentIndex - aIndex;
                if (aIndex < currentIndex) {
                    maxCount = Math.min(aMaxCount, countToGet);
                } else {
                    return;
                }
                break;
            case 1:
                maxCount = Math.min(aMaxCount, queue.length - aIndex + currentIndex);
                break;
            default:
        }

        if (maxCount == 0 || maxCount > MAX_RETRIEVE_COUNT) {
            maxCount = MAX_RETRIEVE_COUNT;
        }

        if (maxCount > currentIndex) {
            handleInternal(objHashIntSet, queue.length - (maxCount - currentIndex), queue.length, bufferTime, handlerConsumer);
            handleInternal(objHashIntSet, 0, currentIndex, bufferTime, handlerConsumer);
        } else {
            handleInternal(objHashIntSet, currentIndex - maxCount, currentIndex, bufferTime, handlerConsumer);
        }
    }

    /**
     * get xlog data and invoke consumer within the range
     *
     * @param from
     * @param to
     * @param bufferTime
     * @param handlerConsumer
     */
    private void handleInternal(IntSet objHashIntSet, int from, int to, long bufferTime, Consumer<XLogPackWrapper> handlerConsumer) {
        long now = server.getCurrentTime();
        for (int i = from; i < to; i++) {
            XLogPackWrapper packWrapper = queue[i];

            //keep bufferTime to consider dictionary info delaying
            if (now - bufferTime < packWrapper.getPack().endTime) {
                break;
            }

            //filter objHash
            if (objHashIntSet != null && objHashIntSet.size() > 0 && !objHashIntSet.contains(packWrapper.getPack().objHash)) {
                continue;
            }
            handlerConsumer.accept(packWrapper);
        }
    }
}

File: scouter.server/src/main/scala/scouter/server/core/cache/XLogLoopCache.scala
/*
*  Copyright 2015 the original author or authors. 
 *  @https://github.com/scouter-project/scouter
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"); 
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License. 
 *
 */

package scouter.server.core.cache;

import java.util.ArrayList;
import java.util.List;

import scouter.util.IntSet;

class XLogLoopCache[V](capacity: Int) {

    val queue = new Array[Any](capacity);
    val objHashTable = new Array[Int](capacity);
    val elapsed = new Array[Int](capacity);
    val error = new Array[Boolean](capacity);

    var index = 0;
    var loop = 0L;

    def put(objHash: Int, time: Int, err: Boolean, record: V) {
        queue.synchronized {
            queue(index) = record;
            objHashTable(index) = objHash;
            elapsed(index) = time;
            error(index) = err;
            index += 1;
            if (index >= queue.length) {
                loop += 1;
                index = 0;
            }
        }
    }

    def getList(start_loop: Long, start_index: Int, time: Int): List[V] = {
        val end_index = this.index;
        val end_loop = this.loop;

        val buff = new ArrayList[V](queue.length);
        (end_loop - start_loop) match {
            case 0 =>
                if (start_index < end_index) {
                    copy(buff, start_index, end_index, time);
                }
            case 1 =>
                if (start_index <= end_index) {
                    copy(buff, end_index, queue.length, time);
                    copy(buff, 0, end_index, time);
                } else {
                    copy(buff, start_index, queue.length, time);
                    copy(buff, 0, end_index, time);
                }
            case _ =>
                copy(buff, end_index, queue.length, time);
                copy(buff, 0, end_index, time);
        }
        return buff;
    }

    def getList(objHashSet: IntSet, start_loop: Long, start_index: Int, time: Int): List[V] = {
        val end_index = this.index;
        val end_loop = this.loop;

        val buff = new ArrayList[V](queue.length);
        (end_loop - start_loop) match {
            case 0 =>
                if (start_index < end_index) {
                    copy(objHashSet, buff, start_index, end_index, time);
                }
            case 1 =>
                if (start_index <= end_index) {
                    copy(objHashSet, buff, end_index, queue.length, time);
                    copy(objHashSet, buff, 0, end_index, time);
                } else {
                    copy(objHashSet, buff, start_index, queue.length, time);
                    copy(objHashSet, buff, 0, end_index, time);
                }
            case _ =>
                copy(objHashSet, buff, end_index, queue.length, time);
                copy(objHashSet, buff, 0, end_index, time);
        }
        return buff;
    }

    def getListWithinCount(start_loop: Long, start_index: Int, _count: Int): List[V] = {
        var  count = _count
        (this.loop - start_loop) match {
            case 0 =>
                val gap = this.index - start_index
                if(gap > 0) {
                    count = Math.min(count, gap)
                }
            case 1 =>
                count = Math.min(count, queue.length - start_index + this.index)
            case _ =>
        }

        val buff = new ArrayList[V](count)
        if(count > this.index) {
            copy(buff, queue.length-(count-this.index), queue.length)
            copy(buff, 0, this.index)
        } else {
            copy(buff, this.index - count, this.index)
        }

        return buff
    }

    def getListWithinCount(objHashSet: IntSet, start_loop: Long, start_index: Int, _count: Int): List[V] = {
        var  count = _count
        (this.loop - start_loop) match {
            case 0 =>
                val gap = this.index - start_index
                if(gap > 0) {
                    count = Math.min(count, gap)
                }
            case 1 =>
                count = Math.min(count, queue.length - start_index + this.index)
            case _ =>
        }

        val buff = new ArrayList[V](count)
        if(count > this.index) {
            copy(objHashSet, buff, queue.length-(count-this.index), queue.length)
            copy(objHashSet, buff, 0, this.index)
        } else {
            copy(objHashSet, buff, this.index - count, this.index)
        }

        return buff
    }

    private def copy(objHashSet: IntSet, buff: List[V], _from: Int, _to: Int, time: Int) {
        var i = _from
        while (i < _to) {
            if (objHashSet == null || objHashSet.contains(objHashTable(i))) {
                if (queue(i) != null && (elapsed(i) >= time || error(i) == true)) {
                    buff.add(queue(i).asInstanceOf[V]);
                }
            }
            i += 1
        }
    }

    def copy(buff: List[V], _from: Int, _to: Int, time: Int) {
        var i = _from
        while (i < _to) {
            if (queue(i) != null && (elapsed(i) >= time || error(i) == true)) {
                buff.add(queue(i).asInstanceOf[V]);
            }
            i += 1
        }
    }

    def copy(buff: List[V], _from: Int, _to: Int) {
        var i = _from
        while (i < _to) {
            if (queue(i) != null) {
                buff.add(queue(i).asInstanceOf[V]);
            }
            i += 1
        }
    }

    private def copy(objHashSet: IntSet, buff: List[V], _from: Int, _to: Int) {
        (_from until _to) foreach { i =>
            if (objHashSet == null || objHashSet.contains(objHashTable(i))) {
                if (queue(i) != null) {
                    buff.add(queue(i).asInstanceOf[V]);
                }
            }
        }
    }

    def get(last_loop: Long, last_index: Int, time: Int): CacheOut[V] = {
        val d = new CacheOut[V]();
        d.data = getList(last_loop, last_index, time);
        d.loop = this.loop;
        d.index = this.index;
        return d;
    }

    def get(objHashSet: IntSet, last_loop: Long, last_index: Int, time: Int): CacheOut[V] = {
        val d = new CacheOut[V]();
        d.data = getList(objHashSet, last_loop, last_index, time);
        d.loop = this.loop;
        d.index = this.index;
        return d;
    }

    def getWithinCount(last_loop: Long, last_index: Int, count: Int): CacheOut[V] = {
        val d = new CacheOut[V]();
        d.data = getListWithinCount(last_loop, last_index, count);
        d.loop = this.loop;
        d.index = this.index;
        return d;
    }

    def getWithinCount(objHashSet: IntSet, last_loop: Long, last_index: Int, count: Int): CacheOut[V] = {
        val d = new CacheOut[V]();
        d.data = getListWithinCount(objHashSet, last_loop, last_index, count);
        d.loop = this.loop;
        d.index = this.index;
        return d;
    }

}